posts/inside-rust/2020-03-02-recent-future-pattern-matching-improvements.md

```diff
@@ -0,0 +1,234 @@
+                    *span,
+                    "`...` must be the last argument of a C-variadic function",
+                );
+            }
+        }
+    }
+}
+```
+
+Here, `ps @ ..` will bind the initial elements of the slice to `ps` and ignore the last element.
+
+After more than 7 years of baking in nightly, with many twists and turns, subslice patterns will finally be stable. To get here, we've had to redesign the feature, plug soundness holes in the borrow checker, and substantially refactor the exhaustiveness checker. For more on how we got here, [read the stabilization report][pr_subslice], [Thomas Hartmann's blog post][thomas_subslice], and stay tuned for the 1.42.0 release announcement  on the 12th of March.
+
+[thomas_subslice]: https://thomashartmann.dev/blog/feature(slice_patterns)/
+
+## Nested OR-patterns
+
+[tracking_or_pats]: https://github.com/rust-lang/rust/issues/54883
+
+When pattern matching on an `enum`, the logic for some of the variants may be exactly the same. To avoid repeating ourselves, the `|` separator in `match`, `if let`, or `while let` expressions can be used to say that the branch should be taken if any of the `|`-separated patterns match. For example, we may write:
+
+```rust
+// Any local node that may call something in its body block should be explored.
+fn should_explore(tcx: TyCtxt<'_>, hir_id: hir::HirId) -> bool {
+    match tcx.hir().find(hir_id) {
+        Some(Node::Item(..))
+        | Some(Node::ImplItem(..))
+        | Some(Node::ForeignItem(..))
+        | Some(Node::TraitItem(..))
+        | Some(Node::Variant(..))
+        | Some(Node::AnonConst(..))
+        | Some(Node::Pat(..)) => true,
+        _ => false,
+    }
+}
+```
+
+This is serviceable, but `Some(_)` is still repeated several times. With [`#![feature(or_patterns)]`][tracking_or_pats], which recently became usable on nightly, this repetition can be avoided:
+
+```rust
+// Any local node that may call something in its body block should be explored.
+fn should_explore(tcx: TyCtxt<'_>, hir_id: hir::HirId) -> bool {
+    match tcx.hir().find(hir_id) {
+        Some(
+            Node::Item(..)
+            | Node::ImplItem(..)
+            | Node::ForeignItem(..)
+            | Node::TraitItem(..)
+            | Node::Variant(..)
+            | Node::AnonConst(..)
+            | Node::Pat(..),
+        ) => true,
+        _ => false,
+    }
+}
+```
+
+Previously, when using `|` in a `match` expression, the `|` syntax was part of `match` itelf. With `or_patterns`, this is now part of patterns themselves, so you can nest OR-patterns arbitrarily, and use them in `let` statements too:
+
+```rust
+let Ok(x) | Err(x) = foo();
+```
+
+An OR-pattern covers the *union* of all the `|`-ed ("or-ed") patterns. To ensure that whatever alternative matched, all bindings are consistent and initialized, each or-ed pattern must include the exact same set of bindings, with the same types, and the same binding modes.
```

Comment:
Why do we need the same binding modes? I would've expected e.g. it be fine to have `Ok(ref x) | Err(x)` with `Result<u32, &u32>` producing `x: &u32`.

Is this just a match/borrow check limitation?

Reply:
It's possible that there's no good reason for this limitation (which is encoded in librustc_resolve) to exist. I'm not exactly sure why it does, so my answer is the non-answer "this is how it has always been" ;)

```
error[E0409]: variable `x` is bound in inconsistent ways within the same match arm
 --> src/main.rs:2:25
  |
2 |     let Ok(ref x) | Err(x) = Ok(0);
  |                -        ^ bound in different ways
  |                |
  |                first binding
```
Maybe @petrochenkov or @matthewjasper knows why.

(Note to self: the diagnostic should not mention "match arm".)

Reply:
The compiler doesn't currently track binding modes separately for each binding. Since `Ok(x) | Err(mut x)` has to be an error, and `Ok(ref x) | Err(&ref x)` is allowed, there isn't much reason to change this.